### PR TITLE
Added a marker for the current position of the confocal scan

### PR DESCRIPTION
**Pull request by NnguyenHTommy**
Added a marker for the current position of the confocal scan done in: src/applications/piezoscan.py. A marker was also added for the current position of the cursor. Tested and confirmed it works on the SiV microscope in QDL.

### CODE REVIEW

File: src/applications/piezoscan.py [hunk 1]
```diff
@@ -98,6 +99,12 @@ def update(self, model):
                                                                    model.xmax + model.step_size,
                                                                    model.current_y + model.step_size,
                                                                    model.ymin])
+
```
Comment (gadamc) [inline src/applications/piezoscan.py:102]: Do we need / want to have the ability to mark a position on the scan "during" the scan?  When I click during a scan (in random data test mode on local Mac) I get sometime get errors on the console.

```
Traceback (most recent call last):
  File "/Users/adam/miniconda3/envs/qt3utilsdevpy312/lib/python3.12/site-packages/matplotlib/cbook.py", line 298, in process
    func(*args, **kwargs)
  File "/Users/adam/miniconda3/envs/qt3utilsdevpy312/lib/python3.12/site-packages/matplotlib/backend_bases.py", line 2972, in mouse_move
    self.set_message(self._mouse_event_to_message(event))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/adam/miniconda3/envs/qt3utilsdevpy312/lib/python3.12/site-packages/matplotlib/backend_bases.py", line 2957, in _mouse_event_to_message
    artists = [a for a in event.inaxes._mouseover_set
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: OrderedDict mutated during iteration
```

I think this is probably a race-condition. The `update` function is being called by two different threads: the main thread and the "scan thread". 

This might be fixed if we get rid of the for-loop around self.ax.lines.

File: src/applications/piezoscan.py [hunk 2]
```diff
@@ -98,6 +99,12 @@ def update(self, model):
                                                                    model.xmax + model.step_size,
                                                                    model.current_y + model.step_size,
                                                                    model.ymin])
+        
+        for line in self.ax.lines:
+                if line.get_label() == 'pos':
+                    line.set_marker('')
+        self.ax.plot(self.circle_pos[0],  self.circle_pos[1], 'ro', label = 'pos')
```
Comment (gadamc) [inline src/applications/piezoscan.py:106]: I think it might be better to not run this loop every time.  The `plot` function returns a line2d object (https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.plot.html).  Something like this might be better


```
if self.circle_pos is not None:
    if self.position_line2d:
        self.position_line2d.data(self.circle_pos)
    else:
        self.position_line2d = self.ax.plot(self.circle_pos[0], self.circle_pos[1], 'ro', label='pos')
```

Alternatively, if the ScanImage object retains the "position_line2d" handle, then the xy position of that object can be updated without calling the ScanImage.update function.

File: src/applications/piezoscan.py [hunk 3]
```diff
@@ -117,9 +124,16 @@ def set_onclick_callback(self, f):
...
     def onclick(self, event):
         if event.inaxes is self.ax:
-            #todo: draw a circle around clicked point? Maybe with a high alpha, so that its faint
             self.onclick_callback(event)
 
+            # draws a x for clicked point
+            for line in self.ax.lines:
+                if line.get_label() == 'pointer':
+                    line.set_marker('')
+            self.ax.plot(event.xdata, event.ydata,'yx', label='pointer')
```
Comment (gadamc) [inline src/applications/piezoscan.py:133]: like above, it's probably going to be better to retain the line2D object returned by ax.plot. Then we can eliminate this for-loop. Also to note, this might be memory-leak because every time a user clicks on the screen it allocates a new chunk of memory -- it creates a new line2D object every time.  Unless Python is cleaning up for us in the background.

File: src/applications/piezoscan.py [hunk 4]
```diff
@@ -339,8 +353,13 @@ def go_to_position(self, event = None):
...
             print(f'stage_controller would have moved to x,y = {self.view.sidepanel.go_to_x_position_text.get():.2f}, {self.view.sidepanel.go_to_y_position_text.get():.2f}')
-        self.optimized_position['x'] = self.view.sidepanel.go_to_x_position_text.get()
-        self.optimized_position['y'] = self.view.sidepanel.go_to_y_position_text.get()
+        x, y  = self.view.sidepanel.go_to_x_position_text.get(),  self.view.sidepanel.go_to_y_position_text.get()
+        self.optimized_position['x'] = x
+        self.optimized_position['y']  =y
+        self.view.scan_view.circle_pos = [x,y]
+        if len(self.counter_scanner.scanned_count_rate) > 0:
+            self.view.scan_view.update(self.counter_scanner)
+            self.view.canvas.draw()
```
Comment (gadamc) [inline src/applications/piezoscan.py:363]: If the ScanImage object retains handles to the "pos" and "pointer" object, we might be able to update the plot by simply updating the data within those object. But I don't know... would have to test and see.

File: src/applications/piezoscan.py [hunk 5]
```diff
@@ -339,8 +353,13 @@ def go_to_position(self, event = None):
             self.counter_scanner.stage_controller.go_to_position(x = self.view.sidepanel.go_to_x_position_text.get(), y = self.view.sidepanel.go_to_y_position_text.get())
         else:
             print(f'stage_controller would have moved to x,y = {self.view.sidepanel.go_to_x_position_text.get():.2f}, {self.view.sidepanel.go_to_y_position_text.get():.2f}')
-        self.optimized_position['x'] = self.view.sidepanel.go_to_x_position_text.get()
-        self.optimized_position['y'] = self.view.sidepanel.go_to_y_position_text.get()
+        x, y  = self.view.sidepanel.go_to_x_position_text.get(),  self.view.sidepanel.go_to_y_position_text.get()
```
Comment (gadamc) [inline src/applications/piezoscan.py:356]: Why was this changed from __init__ to here?  It could certainly be the better place to do this, so I am asking out of curiosity.

File: src/applications/piezoscan.py (hunk 3 above)
Comment (gadamc) [reply src/applications/piezoscan.py:133]: ps: Following the same line of thought, I wonder if line 98 (above) is essentially a memory leak as well -- instead of calling self.ax.imshow and assigning the output to a new handle (self.artist) every time we call 'update', should we instead update the data inside of the self.artist object?  But maybe python is cleaning garbage for us. Would have to profile this.

**Review by gadamc — CHANGES_REQUESTED** [review]
Unfortunately, this seems to break the ability to zoom in on a region of a scan. After scanning, if you click on the 'magnifying glass' at the bottom of the panel, you should be able to drag a square area in the scan in order to zoom in. With this update, that functionality has been blocked.

Also, there are a few other things I think should be done differently, such as retaining a handle to the Line2D object that stores the position of the circle/x.  See below.

File: src/applications/piezoscan.py (hunk 1 above)
Comment (NnguyenHTommy) [reply src/applications/piezoscan.py:102]: I have never gotten this error when clicking during a scan. I think the "feature" is somewhat nice however as I can look at positions and see the counts while the scan is running without having to wait for the scan to finish.

File: src/applications/piezoscan.py (hunk 4 above)
Comment (NnguyenHTommy) [reply src/applications/piezoscan.py:363]: done!

File: src/applications/piezoscan.py (hunk 5 above)
Comment (NnguyenHTommy) [reply src/applications/piezoscan.py:356]: I did it this way since I needed the x,y position for a future use. This prevents 2 calls to the get function which sacrifices speed for memory as I store the x,y as variables.

File: src/applications/piezoscan.py [hunk 6]
```diff
@@ -85,6 +85,7 @@ def __init__(self, mplcolormap = 'gray'):
         self.ax.set_xlabel('x position (um)')
         self.ax.set_ylabel('y position (um)')
         self.log_data = False
+        self.circle_pos = [0, 0]
```
[the remark this answers is not in the document]
Comment (NnguyenHTommy) [reply src/applications/piezoscan.py]: The zoom features works on the SiV microscope. Took out the self.circle_pos handle. Also got rid of showing the red dot during the scan. This was done by using the Line2D object for the markers.

File: src/applications/piezoscan.py (hunk 2 above)
Comment (NnguyenHTommy) [reply src/applications/piezoscan.py]: Done

File: src/applications/piezoscan.py (hunk 3 above)
Comment (NnguyenHTommy) [reply src/applications/piezoscan.py]: Done. no more for loops are used

**Review by gadamc — COMMENTED** [review]
Hi @NnguyenHTommy  I added a new commit where I put the logic of updating the markers in the ScanImage class. 

Can you verify that this works on your system? Then I think it's ready to merge.

**Comment by gadamc** [comment]
In addition to testing on your microscope to be sure,  there are a couple of things we could add
1) the values of the current x/y position to the right panel. currently the "position" in the panel only displays the yellow x pointer
2) add a button to "update current position" -- which reads the voltages to update the red circle. This would be useful in the case where a user updates the position of the laser using something other than qt3utils. It's possible to adjust the laser position by directly turning knobs on the Jena amplifier as well as adjust the position with the 'qt3pizeo' application.

**Review by NnguyenHTommy — COMMENTED** [review]
Tested all the changes on the SiV microscope and they work

**Review by gadamc — APPROVED** [review]
